### ISSUE TITLE
Question with Zero2 Optimization in LoraPro

### ISSUE THREAD

**Issue by LZY-the-boys** [mrflogs/LoRA-Pro#6]
Hi, I find that lorapro zero2 optimization have one strange behavior:
```
            for p, n in zip(group["params"], group["names"]):

                if p.grad is None:
```
zero2 will compress the group['param'] into one tensor,  however, the group['names'] remains to be all the key of model state dict. 
so in essence, this for loop only extract first item of group["params"] (typically have no lora, just embedding), and donnot use the optimizer algorithm lorapro use, does this behavior is expected? Or I misunderstand something? 

the zero2 version I use is `deepspeed==0.15.4` and the other package version is strictly follows the lorapro repo

**Comment by mrflogs** [comment]
We have verified this issue. I suspect that you might have used print to check the output. It appears to be just a display issue, as the combination of transformers and DeepSpeed seems unable to properly show the optimization process here. In a single GPU setup, the full optimization process is displayed correctly. Similarly, when using standard LoRA with torch.optim.AdamW, printing the parameters also shows a similar behavior.

**Comment by LZY-the-boys** [comment]
Thank you for your response!

I have observed that the current optimization process may not be properly invoking the LoRA-Pro specific optimizations. Specifically:

-  When using ZeRO-2, I noticed through debugging with pdb.set_trace() that the solve_sylvester function is not being called. Because the solve_sylvester related logic needs group['name'] to have `LoRA` word. However,  in zero2, the param is a flattened parameter vector, so the `zip(param, name)` appears to fall back to a standard AdamW optimizer applied to ** in a single pass.**
- This behavior becomes particularly noticeable when ZeRO-2 is disabled. Without ZeRO-2, the code execution significantly slows down because it traverses all parameters multiple times, repeatedly invoking solve_sylvester. For models like LLaMA-2, this can result in the training process being up to 60 times slower. Given that ZeRO-2 primarily partitions gradients, such a substantial performance difference seems strange.

I would greatly appreciate any assistance or insights you could provide on this matter.

Best regards,

**Comment by mrflogs** [comment]
We have resolved this issue, and the results have been re-run and validated accordingly.

In LoRA-Pro, to ensure compatibility with DeepSpeed, we've integrated the Adam optimization process directly into DeepSpeed (in [DeepSpeed-0.15.1/deepspeed/runtime/zero/stage_1_and_2.py](https://github.com/mrflogs/LoRA-Pro/blob/main/DeepSpeed-0.15.1/deepspeed/runtime/zero/stage_1_and_2.py#L1828)). Therefore, in the TrainingArguments, you need to set the optimizer to "sgd" to prevent parameters from being updated twice.
